Adding StreamableHttpServerTransportProvider class and unit tests

ZachGerman opened the pull request:
<!-- Provide a brief summary of your changes -->
Still missing:
- Origin header validation (ongoing discussions about spec for that in Discord)
- Backward-compatible endpoint combo

## Motivation and Context
Trying to reach spec parity with TS and Python for Java. Will continue working on other aspects of this.

## How Has This Been Tested?
Unit tests and integ tests using a partially-complete sHTTP client transport class, but I didn't include the integ test file as the client isn't complete.

## Breaking Changes
N/A

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
This is part of a larger commit seen here: https://github.com/modelcontextprotocol/java-sdk/pull/289

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+
+		private String mcpEndpoint;
+
+		private Supplier<String> sessionIdProvider = () -> UUID.randomUUID().toString();
```
Should change the value here to `DEFAULT_SESSION_ID_PROVIDER`. Will address.

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+		try {
+			// Read request body
+			StringBuilder body = new StringBuilder();
+			try (BufferedReader reader = request.getReader()) {
```
Looks like we can use Servlet 3.1's non-blocking support. I understand this implementation is using Jakarta Servlet 6.1, so we can use `request.getInputStream().setReadListener()`

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+	/**
+	 * Implementation of McpServerTransport for Streamable HTTP sessions.
+	 */
+	private class StreamableHttpServerTransport implements McpServerTransport {
```
I am guessing for Streamable HTTP we need a new abstraction, a `McpStreamableServerTransport` that has a different set of methods allowing opening a new stream and staying with the `McpServerTransport` in case of simple (sessionless) servers.

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+			return;
+		}
+
+		// Client accepts SSE since we've validated the Accept header contains
```
See my other comment for `StreamableHttpServerTransport`. We need to distinguish on a per message basis whether we are dealing with a new stream (upgrade to SSE) or a simple response and immediate flush. I imagine this could be handled with a new type of request handler (a new `McpServerSession` type or a new method) which returns a complex type that signifies whether the response is done (Mono) or will stream (Flux) and based on that we can decide what to do. The concept of a transport session from the client implementation should be introduced here as well.

The hierarchy I imagine is as follows:

```
StreamableHttpServerTransportProvider (1)
  StreamableHttpSession (1 per client initiated either with GET or POST)
    StreamableHttpSseStream (N - as many as the client initiates according to the server's upgrade policy and bound via session ID)
      StreamableHttpServerTransport (1 per stream) - this should route generic notifications and requests to one of the open SSE streams, preferably the one opened with GET); notifications, requests and responses tied to a POST should be handled differently (consumed from a Flux), via the containing StreamableHttpSseStream
```

The names are negotiable of course, I just wanted to bring some order to my thoughts.

ZachGerman replied:
SSE upgrade logic is now based on the return type of the tool call.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+		// Subscribe to the SSE stream and write events to the response
+		sseStream.getEventFlux().doOnNext(event -> {
+			try {
+				if (event.id() != null) {
```
Yes. Null event IDs are used to control receiver behavior when reconnecting (no [`Last-Event-ID`](https://html.spec.whatwg.org/multipage/server-sent-events.html#last-event-id) header would be sent in the event of a reconnection being attempted). We should typically set JSON-RPC response event IDs to `null` for this reason, now that you mention it.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+
+		// Log all headers for debugging
+		Enumeration<String> headerNames = request.getHeaderNames();
+		while (headerNames.hasMoreElements()) {
```
Forgot I added an `extractHeaders` method (returns header map) for debugging prior, so just using that and logging the request URI and header map in one line as suggested 👍.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+		}).doOnError(e -> {
+			logger.error("Error in SSE stream: {}", e.getMessage());
+			asyncContext.complete();
+		}).subscribe();
```
TIL about Reactor Context. 😅 Thanks!

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+			}
+
+			// Validate session ID for non-initialize requests
+			if (!isInitializeRequest && sessionId == null) {
```
We can probably just add a single stream session for GET requests into the StreamableHttpSession class. I can add that to my list today.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+			}
+
+			// Handle the message
+			session.handle(message).block(); // Block for servlet compatibility
```
Now writing headers as soon as they're known.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+				AsyncContext asyncContext = request.startAsync();
+				asyncContext.setTimeout(0); // No timeout
+
+				StreamableHttpSseStream sseStream = getOrCreateSseStream(sessionId);
```
Same comment:
We can probably just add a single stream session for GET requests into the StreamableHttpSession class. I can add that to my list today.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,765 @@
+		// Create or get SSE stream for this session
+		StreamableHttpSseStream sseStream = getOrCreateSseStream(sessionId);
+		if (lastEventId != null) {
+			sseStream.replayEventsAfter(lastEventId);
```
Either client or server can close it at any time. Should we add a timer of sorts to hold the stream open and have it default to end right after stream completes?

ZachGerman commented:
Today I'm targeting origin header validation and moving the dedicated GET stream to the StreamableHttpSession class, then adding an integ test for GET on /mcp to start the listening stream.
After that, I believe we should have all core functionality except sessionless and proper SSE response upgrade logic.

chemicL reviewed (commented):
Added a few more comments.

ZachGerman commented:
@sivankri: Changing the related logic after my meeting with Dariusz this morning, so your issue goes away with the new response type mapping.
@tzolov: It's related, but I don't think it's overlapping as I'm just adding the ability to set a list of allowed origins to the server transport provider and enforcing it if it's set.

ZachGerman commented:
> merge commit [6c4830b](https://github.com/modelcontextprotocol/java-sdk/commit/6c4830bf5362ca71887736c357a3cd794fa86c1a) is bad

Oops! Thanks for the heads up! Will fix it today!

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/HttpServletSseServerTransportProvider.java:
```diff
@@ -215,8 +215,8 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
-		HttpServletMcpSessionTransport sessionTransport = new HttpServletMcpSessionTransport(sessionId, asyncContext,
-				writer);
+		HttpServletMcpSessionTransport sessionTransport = new HttpServletMcpSessionTransport(
+				McpServerSession.LISTENING_TRANSPORT, asyncContext, writer);
```
This was unnecessary and being reverted.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/server/McpAsyncStreamableHttpServer.java:
Is this file worth keeping?

sivankri replied:
We should try to leverage McpAsyncServer class with isStreamableHttp boolean property without creating a new class.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/spec/McpServerSession.java:
```diff
@@ -170,13 +243,39 @@ public Mono<Void> handle(McpSchema.JSONRPCMessage message) {
+					}
+					else {
+						logger.error("Invalid request ID: {}", request.id());
+						return Mono.empty();
```
Changed this to:
`return Mono.error(new RuntimeException("Invalid request ID: " + String.valueOf(request.id())));
`

sivankri commented on mcp/src/main/java/io/modelcontextprotocol/server/McpAsyncServer.java:
```diff
@@ -119,7 +123,7 @@ public class McpAsyncServer {
 	McpAsyncServer(McpServerTransportProvider mcpTransportProvider, ObjectMapper objectMapper,
 			McpServerFeatures.Async features, Duration requestTimeout,
-			McpUriTemplateManagerFactory uriTemplateManagerFactory) {
+			McpUriTemplateManagerFactory uriTemplateManagerFactory, boolean isStreamableHttp) {
```
This constructor is called only from overloaded methods present in this class. isStreamableHttp cannot be set to true currently. Please check.

sivankri commented:
@ZachGerman I tried running my mcp remote server against the mcp sdk generated from your branch.  When i used the MCP inspector tool  by setting query parameter transport=streamable-http, i could not get the connection established. Can you please help me with this?

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/spec/McpServerSession.java:
```diff
@@ -203,8 +334,10 @@ private Mono<McpSchema.JSONRPCResponse> handleIncomingRequest(McpSchema.JSONRPCR
-						new TypeReference<McpSchema.InitializeRequest>() {
+				McpSchema.InitializeRequest initializeRequest = transports.isEmpty() ? listeningTransport
+					.unmarshalFrom(request.params(), new TypeReference<McpSchema.InitializeRequest>() {
+					}) : transports.get(request.id())
```
Will push this change at end of day with the update to route client-request-driven server-notifications/requests to the request's response stream if available.

ZachGerman replied:
Didn't get the routing finished yesterday (wanted to enforce `id` rules) but I pushed this change with the `id` changes.
Thanks!

ZachGerman commented:
Removed streamTools in favor of upgrading transport instances to SSE based on McpAsyncServerExchange method usage. 
Also added a more extensive suite of integration tests.
Will speak to @chemicL about adding a config option at the server level to declare default transport type, as I believe he wants that to be SSE for session-bound requests.
Also planning to make event histories some kind of append-only ordered iterable struct, as using `Map.keySet().stream().sorted()` when replaying is a bit gross in hind-sight.

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,893 @@
+
+	@Override
+	public void setSessionFactory(McpServerSession.Factory sessionFactory) {
+		// Required but not used for this implementation
```
We can't implement it this way. This transport needs to be of a different kind.

chemicL commented on mcp/src/main/java/io/modelcontextprotocol/server/transport/StreamableHttpServerTransportProvider.java:
```diff
@@ -0,0 +1,893 @@
+
+					// Only set content type for requests
+					if (message instanceof McpSchema.JSONRPCRequest) {
+						response.setContentType(APPLICATION_JSON);
```
So here you set it to application/json but later you let the user change this to text/event-stream. Unfortunately with other transports it might not be as straightforward and the builders are more type-safe and will forbid rolling back in such a way. That's why I am proposing to keep it simple for now as in another comment and always upgrade to SSE for requests.

ZachGerman commented on mcp-spring/mcp-spring-webflux/src/main/java/io/modelcontextprotocol/client/transport/WebClientStreamableHttpTransport.java:
```diff
@@ -250,11 +247,13 @@ public Mono<Void> sendMessage(McpSchema.JSONRPCMessage message) {
-						// the server to send notifications and requests out-of-band.
-						reconnect(null).contextWrite(sink.contextView()).subscribe();
+					transportSession.markInitialized(response.headers().asHttpHeaders().getFirst("mcp-session-id"));
+					if (response.statusCode().is2xxSuccessful()
+							&& message instanceof JSONRPCNotification notification) {
+						if (notification.method().equals("notifications/initialized")) {
```
After we split the server session class, should we switch them all to mark the session as initialized once they send the `InitializeResult` response? Without that change or making the client wait to get the 200 from its `initialized` notification, we will have clients trying to establish streams prior to initialization being complete.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/spec/McpServerSession.java:
```diff
@@ -89,6 +152,90 @@ public String getId() {
+	 * @param transportId
+	 * @return Map of SseEvent objects, keyed by event ID
+	 */
+	public Map<String, SseEvent> getTransportEventHistory(String transportId) {
```
+1! From previous comment above:
>Also planning to make event histories some kind of append-only ordered iterable struct, as using Map.keySet().stream().sorted() when replaying is a bit gross in hind-sight.

Planning to go with `collections.OrderedDict` unless you have a better recommendation.

ZachGerman commented on mcp/src/main/java/io/modelcontextprotocol/spec/McpServerSession.java:
```diff
@@ -334,6 +578,25 @@ public interface RequestHandler<T> {
+	 * @param <T> the type of the response that is expected as a result of handling the
+	 * request.
+	 */
+	public interface StreamingRequestHandler<T> extends RequestHandler<T> {
```
Yes. Forgot to remove this.
